source/common/ssl/context_config_impl.cc

```diff
@@ -1,17 +1,43 @@
 #include "common/ssl/context_config_impl.h"
 
+#include <memory>
 #include <string>
 
 #include "common/common/assert.h"
 #include "common/config/datasource.h"
 #include "common/config/tls_context_json.h"
 #include "common/protobuf/utility.h"
+#include "common/ssl/tls_certificate_config_impl.h"
 
 #include "openssl/ssl.h"
 
 namespace Envoy {
 namespace Ssl {
 
+namespace {
+
+std::string readConfig(
+    const envoy::api::v2::auth::CommonTlsContext& config, Secret::SecretManager& secret_manager,
+    const std::function<std::string(const envoy::api::v2::auth::TlsCertificate& tls_certificate)>&
+        read_inline_config,
+    const std::function<std::string(const std::shared_ptr<Ssl::TlsCertificateConfigImpl>& secret)>&
+        read_secret) {
+  if (!config.tls_certificates().empty()) {
+    return read_inline_config(config.tls_certificates()[0]);
+  } else if (!config.tls_certificate_sds_secret_configs().empty()) {
+    auto name = config.tls_certificate_sds_secret_configs()[0].name();
+    auto secret = secret_manager.findSecret(Secret::Secret::SecretType::TLS_CERTIFICATE, name);
+    if (!secret) {
+      throw EnvoyException(fmt::format("Static secret is not defined: {}", name));
+    }
+    return read_secret(std::dynamic_pointer_cast<Ssl::TlsCertificateConfigImpl>(secret));
```

Comment:
I'm still really not a fan of this dynamic cast stuff. Is there any real benefit to doing it this way? Can we just have the secret manager manage different types of secrets and have different add/get functions for each of them? They can all derive from a base secret type so that common logic can be done across all secrets at a later time?

With that said, I don't feel super strongly about this. @PiotrSikora any opinions here given what we need to do in the future in this area?

Reply:
@PiotrSikora Could you please update your opinions?

Reply:
slightly +1 to @mattklein123. We can move back to this dynamic cast stuff where the type of secret increases.